### PR TITLE
ISPN-8708 Remote iteration with limited batch sends finished segments

### PR DESCRIPTION
**Pull request by wburns**
too early

* Use KeyWatchingListener for each publisher

https://issues.jboss.org/browse/ISPN-8708

### CODE REVIEW

**Comment by rvansa** [comment]
@wburns I guess it wouldn't be easy to create a reproducer, would it?

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 1]
```diff
@@ -651,109 +662,154 @@ public void close() {
...
+         this.completionListeners = Collections.synchronizedList(new ArrayList<>(4));
+         this.userListener = completionListener;
       }
 
       @Override
       protected S getNext() {
          S next = super.getNext();
          if (next != null) {
-            completionListener.valueIterated(next);
+            for (KeyWatchingCompletionListener keyWatchingCompletionListener : completionListeners) {
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: is it guaranteed that all listeners will be initialized when this gets called?

Comment (wburns) [reply core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Yes, they are always initialized the way the code works.

And actually if this changes it doesn't matter since the listeners only care about entries that were retrieved by their own publisher.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 2]
```diff
@@ -541,11 +534,29 @@ private IdentityPublisherDecorator() { }
                lostSegments.accept((Supplier<PrimitiveIterator.OfInt>) ourSegments::iterator);
             }
          });
-         return decorateBeforeReturn(convertedPublisher);
+         return iteratorOperation.handlePublisher(convertedPublisher, keyConsumer);
       }
+   }
+
+   private class RehashPublisherDecorator<S> extends AbstractRehashPublisherDecorator<S> {
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Sure, I can do that. Unfortunately these classes will only make sense in the scope of this class, but for clarity I can move them out.

**Comment by wburns** [comment]
I might be able to create a reproducer that uses a controlled Publisher, let me look into it a bit.

**Comment by danberindei** [comment]
@wburns Which test?

**Comment by wburns** [comment]
The test is https://github.com/infinispan/infinispan/blob/master/client/hotrod-client/src/test/java/org/infinispan/client/hotrod/impl/iteration/MultiServerDistRemoteIteratorTest.java

I could usually reproduce about 40% of the time.

**Comment by wburns** [comment]
I am almost done with a test that always reproduces it, unfortunately it is very verbose :(

**Comment by wburns** [comment]
Updated and added test that failed on previous implementation, but new one passes.

**Comment by rvansa** [comment]
Thanks, Will. I'll ask you, though, to create a `package-info.java` in `o.i.stream.impl` and document the big picture with reference to the concrete classes. The code has grown in complexity and `ClusterStreamManager` javadoc is not exhaustive.

**Comment by wburns** [comment]
Yes I can agree some additional documentation would be helpful.

I am also finding that while using Producer/Subscriber is very flexible it is extremely hard to follow code as you don't know which is which most of the time :(

I was planning on writing a blog post about these changes in a very high level, I will try to get some actual documentation in code at the same time.

Thanks @rvansa

**Comment by rvansa** [comment]
It's not a problem of the reactive APIs but also when e.g. using suppliers. I often had to trace a long chain to find out what is the actual value (e.g. if is given supplier of entry set providing just the local or remote entries etc.). Calling variables `supplier` then does not help much.

**Comment by wburns** [comment]
Yeah I realized this as well. I unfortunately only renamed the _userListener_ in the CompletionRehashPublisherDecorator which helped a little, but unfortunately there are quite a few :)
